homeassistant/components/sonos/media_player.py

```diff
@@ -1281,3 +1338,140 @@ def device_state_attributes(self):
             attributes[ATTR_QUEUE_POSITION] = self.queue_position
 
         return attributes
+
+    async def async_browse_media(self, media_content_type=None, media_content_id=None):
+        """Implement the websocket media browsing helper."""
+        if media_content_type in [None, "library"]:
+            return await self.hass.async_add_executor_job(
+                library_payload, self._media_library
+            )
+
+        payload = {
+            "search_type": media_content_type,
+            "idstring": media_content_id,
+            "full_album_art_uri": True,
+            "max_items": 0,
+        }
+        response = await self.hass.async_add_executor_job(
+            build_item_response, self._media_library, payload
+        )
+        if response is None:
+            raise BrowseError(
+                f"Media not found: {media_content_type} / {media_content_id}"
+            )
+        return response
+
+
+def build_item_response(media_library, payload):
+    """Create response payload for the provided media query."""
+    if payload["search_type"] == TYPE_ALBUM and payload["idstring"].startswith(
+        ("A:GENRE", "A:COMPOSER")
+    ):
+        payload["idstring"] = "A:ALBUMARTIST/" + "/".join(
+            payload["idstring"].split("/")[2:]
+        )
+
+    media = media_library.browse_by_idstring(**payload)
+
+    if media is None:
+        return
+
+    thumbnail = None
+    title = None
+
+    # Fetch album info for titles and thumbnails
+    # Can't be extracted from track info
+    if (
+        payload["search_type"] == TYPE_ALBUM
+        and media[0].item_class == "object.item.audioItem.musicTrack"
+    ):
+        for item in media_library.browse_by_idstring(
```

Comment:
Each item will do requests here to the Sonos API, is that a problem? Should we just cache some data maybe or make sure we only fetch certain results once per building a response?

For example, let's say I am navigating an artist that has 10 albums, we will now fetch 10 times all albums just to pick that one album?

Reply:
I haven't noticed any issues and this is all local. But I'll look into that.

Reply:
But I agree, it is a lot of effort for just fetching album art. The API is a bit cumbersome IMO.

Reply:
```python
from functools import lru_cache

cached_browse = lru_cache(maxsize=None)(media_library.browse_by_idstring)
```

Then pass that browse function around to have things fetch stuff.

Reply:
But only keep that function around while handling the request, not forever, since it doesn't cap the cache size.

Reply:
After discussion on discord we decided against caching for now.